### PR TITLE
Feature/designation field automatic filling

### PR DESCRIPTION
**Pull request by deeonwuli**
### :pushpin: References

* **Issue:** Closes https://app.clickup.com/t/3b501v2

### :art: Screenshots
Screenshot 2022-08-18 at 10 26 23
Screenshot 2022-08-18 at 10 42 03

### CODE REVIEW

File: src/components/steps/general-info/Funders.tsx [hunk 1]
```diff
@@ -27,32 +40,78 @@ const Funders: React.FC<FundersProps> = ({ project, onChange }) => {
...
+                    }}
+                    maxWidth="sm"
+                    fullWidth={true}
                 />
-            </div>
-        </CardContent>
+            )}
+        </>
     );
 };
```
Comment (tokland) [inline src/components/steps/general-info/Funders.tsx:116]: There are some issues:

- Functionally: 
  - Note that when saying "yes" to add the code, it concatenates all the selected funder codes, not the latest added (which I understand is the only one to be added).
  - Problem: Add one funder, say "no". Add another funder, say "yes" -> no code is added. Add another funder, say "yes" -> now all funders are added.

Implementation: The `isDialogOpen/automaticFill/value` infrastructure is too convoluted. Let's do it simpler:
 - We only need one state variable, let's call it `fundersAdded` or something similar, with type `Maybe<Array<{code: string}>>`.  
 - Note that `MultiSelector->onChange` returns all the selected funders, but we are interested only in the newly added. Make a diff between `project.funders` and `selected.` That's the value to set the new state variable.

File: src/components/steps/general-info/Funders.tsx [hunk 2]
```diff
@@ -27,32 +40,78 @@ const Funders: React.FC<FundersProps> = ({ project, onChange }) => {
...
+                    cancelText={i18n.t("No")}
+                    saveText={i18n.t("Yes")}
+                    onCancel={() => {
+                        setAutomaticFill(false);
+                        setDialogOpen(false);
+                    }}
+                    onSave={() => {
+                        setAutomaticFill(true);
+                        if (automaticFill) {
+                            const additionalDesignation = value.reduce(
```
Comment (tokland) [inline src/components/steps/general-info/Funders.tsx:101]: Some kind of `Array.join` should work.

**Review by tokland — CHANGES_REQUESTED** [review]

File: src/components/steps/general-info/Funders.tsx (hunk 1 above)
Comment (deeonwuli) [reply src/components/steps/general-info/Funders.tsx:111]: @tokland i haven't been able to figure out how to use one state variable to fix the problem. i have just pushed something that fixes it but i don't know if it's the best solution.

File: src/components/steps/general-info/Funders.tsx [hunk 3]
```diff
@@ -15,20 +15,20 @@ interface FundersProps {
...
-    const [automaticFill, setAutomaticFill] = React.useState(true);
-    const [value, setValue] = React.useState<
-        {
-            id: string;
-            displayName: string;
-            shortName: string;
-        }[]
-    >([]);
+    const [fundersAdded, setFundersAdded] = React.useState<Funder[]>([]);
+    const [code, setCode] = React.useState<string[]>([]);
```
Comment (tokland) [inline src/components/steps/general-info/Funders.tsx:32]: Note: I had also doubts with this, but after checking with Adrian, we want to set all the selected funder codes, not the last selected. So it's all very simple:

```
    const [isDialogOpen, setDialogOpen] = React.useState(false);
    const [fundersAdded, setFundersAdded] = React.useState<Funder[]>([]);
    const [code, setCode] = React.useState<string[]>([]);
```

We only need the first. The other two are not really state of the component.

```
    const [isDialogOpen, setDialogOpen] = React.useState(false);
```

File: src/components/steps/general-info/Funders.tsx [hunk 4]
```diff
@@ -92,18 +92,13 @@ const Funders: React.FC<FundersProps> = ({ project, onChange }) => {
...
-                        setAutomaticFill(true);
-                        if (automaticFill) {
-                            const additionalDesignation = value.reduce(
-                                (accumulator, { shortName }) =>
-                                    accumulator ? `${accumulator}-${shortName}` : shortName,
-                                ""
-                            );
-                            updateAdditionalDesignationField("additional", additionalDesignation);
+                        if (fundersAdded) {
+                            code.push(fundersAdded[fundersAdded.length - 1].shortName);
```
Comment (tokland) [inline src/components/steps/general-info/Funders.tsx:99]: As stated above, simply use `project.funders` and concatenate codes with join, no need for state.

**Review by tokland — APPROVED** [review]
All good! I've just added a compact() for the codes, as the type designate them as optional.

**Review by adrianq — APPROVED** [review]
Thanks @deeonwuli It works like a charm!

**Pull request merged**